docs/usage/operations.rst

```diff
@@ -396,3 +396,112 @@ Class Definitions
                   name="foo_unique",
               ),
           ]
+
+
+.. py:class:: SaferAlterFieldSetNotNull(model_name: str, name: str)
+
+    Provides a safer way to alter a field to NOT NULL.
+
+    :param model_name: Model name in lowercase without underscores.
+    :type model_name: str
+    :param name: The column name to be set as not null.
+    :type name: str
+    :param field: The field that is being changed.
+    :type field: models.Field
+
+    **Why use this SaferAlterFieldSetNotNull operation?**
+    -----------------------------------------------------
+
+    When using Django's default AlterField operation, the SQL created has the
+    following form:
+
+    .. code-block:: sql
+
+      ALTER TABLE "foo" ALTER COLUMN "bar" SET NOT NULL;
+
+    This statement will acquire an access exclusive lock on the "foo" table
+    while it rescans the table to find potential violations.
+
+    All reads and writes will be blocked.
+
+    This operation leverages Postgres constraints to safely set the column to
+    not null. This operation will trigger the following queries:
+
+    .. code-block:: sql
+
+      -- The below still requires ACCESS EXCLUSIVE lock, but doesn't require a
+      -- full table scan.
+      -- This check will only be applied to new or modified rows, existing rows
+      -- won't be validated because of the NOT VALID clause.
+      ALTER TABLE foo
+      ADD CONSTRAINT bar_not_null
+      CHECK (bar IS NOT NULL) NOT VALID;
+
+      -- The below performs a sequential scan, but without an exclusive lock.
+      -- Concurrent sessions can read/write.
+      -- The operation will require a SHARE UPDATE EXCLUSIVE lock, which will
+      -- block only other schema changes and the VACUUM operation.
+      ALTER TABLE foo VALIDATE CONSTRAINT bar_not_null;
+
+      -- Requires ACCESS EXCLUSIVE LOCK, but bar_not_null proves that there
+      -- is no NULL in this column and a full table scan is not required.
+      -- Therefore, the ALTER TABLE command should be fast.
+      -- If you are not on Postgres <= 11, you should skip this as it will
+      -- take a lot of time, and the current CHECK constraint might be good
+      -- enough.
+      ALTER TABLE foo ALTER COLUMN bar SET NOT NULL;
```

Comment:
Advising "you should skip this" doesn't make much sense, given that `SaferAlterFieldSetNotNull` doesn't provide a way to skip it.

Reply:
copy-pasta fixed in: b3f5693424df4f8f620308eb11c59058fce5911d